libafl/src/corpus/id_manager.rs

```diff
@@ -0,0 +1,124 @@
+//! The [`CorpusIDManager`] is responsible for keeping track of active [`CorpusID`]s. It lets the corpus map a
+//! [`CorpusID`] to its corresponding corpus index to allow for a vector-based corpus storage. It is the only component
+//! that should be able to create new [`CorpusID`]s.
+
+use alloc::vec::Vec;
+use core::fmt::{Display, Error, Formatter};
+use serde::{Deserialize, Serialize};
+
+use crate::bolts::rands::Rand;
+use crate::inputs::Input;
+use crate::state::{HasCorpus, HasRand};
+
+use super::Corpus;
+
+/// Creates a new [`CorpusID`]. [`CorpusID`]s are globally unique for a corpus and monotonically increasing.
+/// They should only ever be created by a [`CorpusIDManager`], everything else must acquire them from one.
+/// [`CorpusID`]s allow us to track a testcase uniquely even when corpora can have testcases be removed or replaced.
+/// Two different testcases should never have the same [`CorpusID`].
+#[derive(Debug, Copy, Clone, Hash, Eq, PartialEq, Ord, PartialOrd, Serialize, Deserialize)]
+pub struct CorpusID {
+    identifier: usize,
+}
+impl Display for CorpusID {
+    fn fmt(&self, f: &mut Formatter) -> Result<(), Error> {
+        write!(f, "CorpusID[#{}]", self.identifier)
+    }
+}
+
+/// A [`CorpusIDManager`] is responsible for keeping track of active [`CorpusID`]s. It creates new ones, ensures that
+/// they are unique, and maps them to their corresponding indices in the corpus. The `active_ids` field
+#[derive(Debug, Default, Clone, Serialize, Deserialize)]
+pub struct CorpusIDManager {
+    next: usize,
+    active_ids_: Vec<CorpusID>,
+}
+
+impl CorpusIDManager {
+    /// Creates a new [`CorpusIDManager`].
+    #[must_use]
+    pub fn new() -> Self {
+        CorpusIDManager::default()
+    }
+
+    /// Get a slice of the currently active [`CorpusID`]s. The length of this slice should always match the `count()` of
+    /// the corpus.
+    #[must_use]
+    pub fn active_ids(&self) -> &[CorpusID] {
+        // should always be sorted, since it only increasing values are ever appended!
+        // debug_assert!(self.active_ids_.is_sorted());
+        &self.active_ids_
+    }
+
+    /// Allocate the next [`CorpusID`] and return it. This returns a [`CorpusID`] with an identifer larger than all
+    /// previously issued [`CorpusID`]s. This new [`CorpusID`] is immediately added to the `active_ids`.
+    pub(super) fn provide_next(&mut self) -> CorpusID {
+        let val = self.next;
+        self.next = val.checked_add(1).unwrap();
```

Comment:
Instead of unwrapping, you can return an Error from this function, then it's recoverable.
This will never happen (usize corpus elements...) but if you're already using checked add, might as well make it recoverable